### DISCUSSION TITLE
Welcome to Key2Joy Discussions!

### DISCUSSION THREAD

**Discussion by luttje** [Key2Joy/Key2Joy#55] [Announcements]
## 👋 Welcome!
We’re using Discussions as a place to connect with other members of our community. GitHub discussions are like a small forum where you make a discussion (a.k.a. thread/topic) to which others can reply.

We hope that you:
  * [Ask questions you’re wondering about.](https://github.com/luttje/Key2Joy/discussions/categories/q-a)
  * Engage with other community members.
  * [Share ideas for features](https://github.com/luttje/Key2Joy/discussions/categories/ideas)
  * [Share something cool you made](https://github.com/luttje/Key2Joy/discussions/categories/show-and-tell)
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

## 🐞 Found a bug?
Please [create an issue under 'Issues'](https://github.com/luttje/Key2Joy/issues) so we can track progress on fixing them. 

Feel free to use Discussions to explore with us whether you've found a bug, or are overlooking something.

## ❤ Thanks for your input!
We appreciate your bug reports, contributions, help or other positive/constructive interactions and feedback. With your help we can make this free and open-source software useful to all who need it.

**Comment by hamasakifanz** [comment]
Hi there :) thanks for this forum and wonderful app.
I'm an Apex legend player. Apex is a game where you can use MnK and Controller as an input, and they work simultaneously at the same time. The thing is, controller has Aim Assist, so I'd like to try to simulate Controller experience on MnK.

I downloaded 0.7 latest version, but inside there's no default.cfg file like in version 0.6 (and they have completely different format). For starter, can you please share a default.cfg file where all controller's buttons are mapped to keyboard keys and mouse? 
Thanks.

**Reply by luttje** [reply]
Hi again!

Can you try the following:
1. Make sure Key2Joy is shut down
2. Rename the Key2Joy directory in your documents folder to `Key2Joy.bak` (creating a backup)
3. Start Key2Joy

Since 0.7 the default profile file is contained inside the .exe. It should extract it automatically upon startup, if no profile is found. (This makes automated tests a bit easier for me)

I take it that if there is an existing file, it does not extract the default profile, which I consider a bug and I'll make an issue for it.

I just tested it and it worked on my machine, but please let me know if it worked for you

**Reply by hamasakifanz** [reply]
It works! Thanks.

**Reply by hamasakifanz** [reply]
![image](https://github.com/luttje/Key2Joy/assets/13132725/b041f50d-ff5a-4c9a-af6d-2e1afedd3a6f)

I tried to use all of the mappings provided in the interface, but somehow I couldnt find the mapping for R2 and L2 buttons which are crucial buttons (blue question marks). 

Secondly, are those buttons with grey question marks mapable? (not so important)

Also I dont know what these 3 button represent in controller: LeftStickAsAnalog, RightStickAsAnalog, DPadAsAnalog.

**Reply by luttje** [reply]
Those R2 and L2 buttons are called 'Triggers'. They have their own 'Trigger' (confusingly named the same):
![image](https://github.com/luttje/Key2Joy/assets/2738114/48110ed7-d1eb-4750-a6e2-4311213d7ce9)

This is separate because, unlike buttons which are either pressed or not, GamePad Triggers can be pressed 'a bit', 'completely' or anything in between.

Those other buttons (marked with gray question marks) cannot currently be mapped to I believe. I'm not sure if they can even be intercepted at all.

And as for these buttons: `LeftStickAsAnalog, RightStickAsAnalog, DPadAsAnalog`. I too have no idea what they represent. I left them in, because I thought they might be useful to someone who knows what they do. The package used internally [SimWinInput](https://github.com/DavidRieman/SimWinInput/blob/main/SimWinGamePad/GamePadControl.cs) exposes them.
However looking at that code of that package again, it seems it doesn't mark those buttons as 'binary' (pressed or not), so I guess I should remove them from the options. I'll make an issue for that right away!

Let me know if any questions remain, happy to help!

**Reply by hamasakifanz** [reply]
Thanks, I tried it and it works. 
I will play around more with the settings. When I tested on https://hardwaretester.com/gamepad , everything looks fine WASD on left stick and mouse movement on right stick, but when I tested in game, somehow a slight movement on my mouse causes the crosshair to move down on top of the left/right direction. Also sometimes W (left stick move up) and D (left stick move right) became unresponsive. I had to press either WD or SD to make it responsive again. Not sure why these 2 issue happened.

**Reply by luttje** [reply]
Hmm that sounds like a horrible experience. I'm firing up some games to test with and see if I can replicate the problem. If not I'll install Apex Legends sometime in the future and test it.

By the way I want you to be aware that this tool **could be detected as a cheating device**. You should be careful, since using it may get you banned.

**Reply by luttje** [reply]
@hamasakifanz I'm experiencing similar problems to what you described. I've made issues for them in #60 and #61. However I can't promise to fix this very soon, I have a lot of projects going on and only so much spare time.

From my testing it does seem that version 0.6.0 functions better with the mouse move mapped to a stick. But it seems the W/A/S/D getting stuck is present in both versions.

Note that to get 0.6.0 to function you'll have to rename your Documents/Key2Joy folder again.

**Reply by hamasakifanz** [reply]
Thanks for the heads up 😅 I'm not using it in real match, just in practice mode.
No worries about the time, I'm just curious if it's possible to have Controller experience with MnK. 

Yeah, yesterday when I tried v0.6 it wasn't this bad, but this 'disable mouse and keyboard' featyre is crucial I think for Apex. 
When I tried v0.6 it seems moving mouse also moves both mouse and joystick. This causes the game to detect keyboard input and disabled aim assist (not 100% sure tho, just my conclusion from a quick 10 minutes observation)

**Reply by luttje** [reply]
You're right that that version 0.6 also sends mouse input to games. Blocking didn't work there.

I've made a quick fix and described the problem and how to use the fix in this tip: #63. It's still a bit weird to perform, but I'll work on improving this in the future (#64). You will need the latest pre-release for it to work: https://github.com/luttje/Key2Joy/releases/tag/0.7.0-alpha.6a3d028804

This still doesn't fix the WASD movement getting stuck sometimes, but I believe I've at least found out what the problem is.

**Reply by matteo1307** [reply]
Hi,

(sorry for my not so perfect english)

 i also play Apex and i have a pretty rough time to set up the Mouse to Joystick config, could you share your Key2Joy settings so i can conpare ?

**Reply by luttje** [reply]
> Hi,
> 
> (sorry for my not so perfect english)
> 
> i also play Apex and i have a pretty rough time to set up the Mouse to Joystick config, could you share your Key2Joy settings so i can conpare ?

Hi, best try an older version:
https://github.com/luttje/Key2Joy/releases/tag/0.6.0

You can find the default config here: https://github.com/luttje/Key2Joy/blob/fe638f3dac8b3fbdf8c692c02364d9ddb845bdb5/Key2Joy/default-profile.k2j.json

**Comment by TwistedSteel2650** [comment]
Hi Luttje.

In your latest pre release how do you map wasd? Can't seem to figure it out.. Seems like the way in your 0.60 is removed.

**Reply by luttje** [reply]
Hi @TwistedSteel2650,

Thanks for your message. Mapping WASD should be similar to in 0.6.0 unless I'm mistaking, or I'm misunderstanding your question.

Here's a screenshot of the default mapping for reference:
![afbeelding](https://github.com/luttje/Key2Joy/assets/2738114/6808dc22-2b25-499e-a8cd-19cd8cb37259)

To get the default mapping:
1. Go to `C:\Users\<your Windows username>\Documents`
2. Make sure Key2Joy isn't running
3. Rename the `Key2Joy` folder to something like `Key2Joy.bak` or `Key2Joy-backup`
4. Start Key2Joy and it will create the default mapping in a new `Key2Joy` folder

Let me know if this helps or if you have more questions.

**Reply by TwistedSteel2650** [reply]
Thanks for the answer. I figured it out.

Is there a solution, so you can map a button to enable the arm mappings?

**Reply by luttje** [reply]
Not yet, however we will add it in the future. This issue tracks that: #69

😊

**Comment by vini4842** [comment]
Hello Luttje.
How are you?
This application has a very promising future!

Sorry for my rust english :/

I tried to perform some tests in Call of Duty, but unfortunately the mouse command causes the crosshair to rotate and does not respond correctly to commands.
Do you have any idea how to resolve this?
In the near future I will be available to help with the translation into PT-BR.

**Reply by luttje** [reply]
Hi @vini4842, I'm alright thanks for asking :)

Thanks for your interest in this software. Your problem in Call of Duty sounds like one of the bugs I haven't been able to solve yet (#61, #62 or perhaps even #60) 

I'm a bit busy with other projects at the moment, so I won't be able to fix the problems described.

**Perhaps an older version works for you? https://github.com/luttje/Key2Joy/releases/tag/0.6.0**

Let me know if it works and thanks for your offer to help with translation in the future (#10). Much appreciated!

**Reply by paulomega50** [reply]
hello @luttje This application has a very promising future! and very potencial, i try in call of duty and crosshair rotate and does not reespond correctly but i use oldest version and work better the aim, but whit a long time play de game lag i dont know why, i know its not easy work in project like this and its a need a lot of time but if your upgrade your app its will be very promising future ! like rewasd

**Reply by luttje** [reply]
Hi @paulomega50 Thanks for your positive message, I really appreciate that. Also thanks for checking out the older version and confirming that it works better, I'll look into the bugs when I have more time (probably july/august this year).
Thanks for your report of the game lag, this is something I'll have to look into then.
Kind regards,

**Reply by paulomega50** [reply]
you are a genius and I admire your work and your application, I have seen a lot of people selling this type of application that does not work, your application is good I am sure that with calm and a little time it will come out perfect, I truly admire your work, greeting from your biggest admirer!

**Comment by Junaid4798** [comment]
my aim go downwards can anyone help me

**Reply by luttje** [reply]
Hi! Which version are you on? The latest versions are having some issues, sometimes similar to what you describe. 

Perhaps you can try an earlier version like 0.6.0?

**Reply by Junaid4798** [reply]
i am using 0.6.0 while playing call of duty warzone aim doesent work properly

**Reply by luttje** [reply]
Thanks for the report and providing more information. Sadly I won't be able to test this, and due to other projects I don't want to dedicate time to fix this problem at the moment. Sorry I can't be of more help here. 

Best of luck!

**Comment by possionthefish1** [comment]
Just wanna say thank you. Easy to setup and works wonderfully. Thanks again.

**Comment by paulomega50** [comment]
Oh that’s is good news my friend, I definitely will try tank you ! You are
amazing!

possionthefish1 ***@***.***> escreveu em seg., 18/11/2024 às
13:41 :

**Comment by Christoph-Heinrich-Clemens** [comment]
What a great tool! Finally a controller emulator that just works. 
If I could have a wish or two it would be multiple mouse and keyboard ids for multiplayer and locking/hiding a specific device (e.g. mouse based Lightguns or Wiimotes).
